cmd/storage-rest-server.go

```diff
@@ -0,0 +1,352 @@
+	filePath := vars[storageRESTFilePath]
+	fileSizeStr := vars[storageRESTLength]
+	fileSize, err := strconv.Atoi(fileSizeStr)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = s.storage.PrepareFile(volume, filePath, int64(fileSize))
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// AppendFileHandler - append to a file.
+func (s *storageRESTServer) AppendFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	if r.ContentLength < 0 {
+		s.writeErrorResponse(w, errInvalidArgument)
+		return
+	}
+
+	buf := make([]byte, r.ContentLength)
+	_, err := io.ReadFull(r.Body, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = s.storage.AppendFile(volume, filePath, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// StatFileHandler - stat a file.
+func (s *storageRESTServer) StatFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	info, err := s.storage.StatFile(volume, filePath)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	defer w.(http.Flusher).Flush()
+	gob.NewEncoder(w).Encode(info)
+}
+
+// ReadAllHandler - read all the contents of a file.
+func (s *storageRESTServer) ReadAllHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	buf, err := s.storage.ReadAll(volume, filePath)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	w.Header().Set("Content-Length", strconv.Itoa(len(buf)))
+	w.Write(buf)
+}
+
+// ReadFileHandler - read section of a file.
+func (s *storageRESTServer) ReadFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+	offset, err := strconv.Atoi(vars[storageRESTOffset])
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	length, err := strconv.Atoi(vars[storageRESTLength])
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	if offset < 0 || length < 0 {
+		s.writeErrorResponse(w, errInvalidArgument)
+		return
+	}
+	var verifier *BitrotVerifier
+	if vars[storageRESTBitrotAlgo] != "" {
+		hashStr := vars[storageRESTBitrotHash]
+		var hash []byte
+		hash, err = hex.DecodeString(hashStr)
+		if err != nil {
+			s.writeErrorResponse(w, err)
+			return
+		}
+		verifier = NewBitrotVerifier(BitrotAlgorithmFromString(vars[storageRESTBitrotAlgo]), hash)
+	}
+	buf := make([]byte, length)
+	_, err = s.storage.ReadFile(volume, filePath, int64(offset), buf, verifier)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	w.Write(buf)
```

Comment:
I think we need to add Content-Length for this call also (it looks like this is needed so the client will know that data transfer is finished and then it can immediately continue to send requests on the same connection)

Reply:
Yes makes sense.